core-to-plc/src/Language/Plutus/CoreToPLC/Types.hs

```diff
@@ -0,0 +1,53 @@
+{-# LANGUAGE ConstraintKinds  #-}
+{-# LANGUAGE FlexibleContexts #-}
+
+module Language.Plutus.CoreToPLC.Types where
+
+import           Language.Plutus.CoreToPLC.Error
+
+import qualified Language.PlutusCore             as PLC
+import           Language.PlutusCore.Quote
+
+import qualified GhcPlugins                      as GHC
+
+import           Control.Monad.Except
+import           Control.Monad.Reader
+import           Control.Monad.State
+
+import qualified Data.List.NonEmpty              as NE
+import qualified Data.Map                        as Map
+import qualified Data.Text                       as T
+
+type PLCExpr = PLC.Term PLC.TyName PLC.Name ()
+type PLCType = PLC.Type PLC.TyName ()
+
+type ConvError = WithContext T.Text (Error ())
+
+type PrimTerms = Map.Map GHC.Name (Quote PLCExpr)
+type PrimTypes = Map.Map GHC.Name (Quote PLCType)
+
+newtype ConversionOptions = ConversionOptions { coCheckValueRestriction :: Bool }
+
+type ConvertingContext = (ConversionOptions, GHC.DynFlags, PrimTerms, PrimTypes, ScopeStack)
+
+data EvalState a = Done | InProgress a
+type TypeDefs = Map.Map GHC.Name (EvalState (PLC.TyName ()))
+newtype ConvertingState = ConvertingState { csTypeDefs :: TypeDefs }
+
+-- See Note [Scopes]
+type Converting m = (Monad m, MonadError ConvError m, MonadQuote m, MonadReader ConvertingContext m, MonadState ConvertingState m)
+
+{- Note [Scopes]
+We need a notion of scope, because we have to make sure that if we convert a GHC
+Var into a variable, then we always convert it into the same variable, while also making
+sure that if we encounter multiple things with the same name we produce fresh variables
+appropriately.
```

Comment:
I thought, the purpose of uniques is to avoid tracking scopes? Or do we not have access to GHC uniques?

Reply:
We do have GHC uniques, but we'd need to map them into our uniques, which is tricky, or just treat them as part of the name, which is ugly.

It's easy enough to carry scopes, it makes me more sure that it's correct without having to worry about the global uniqueness property, and it gives us better names.

Reply:
> It's easy enough to carry scopes, it makes me more sure that it's correct without having to worry about the global uniqueness property, and it gives us better names.

I agree with this view in general.

Reply:
TBH, I think, it would be ok to treat them as part of the name. This is what you usually do during code generation in a compiler.